### ISSUE TITLE
There was a problem with spectral. Error running Nimma when running lint-openapi program

### ISSUE THREAD

**Issue by loucadufault** [IBM/openapi-validator#373]
I am trying to lint my `openapi.json` spec file.

I have installed the package as a `devDependency` for my Node project.

I have followed the installation guide and installed the `ibm-openapi-validator` package using:

```
npm i -D ibm-openapi-validator
```

which installs version 0.53.0.

Upon running the validator with

```
npx lint-openapi openapi.json --debug
```

It gives the error:

```
[Error] There was a problem with spectral.
Error running Nimma

Error: Error running Nimma
    at Scope.destroy (/Users/me/Projects/proj/node_modules/nimma/dist/legacy/cjs/runtime/scope.js:137:13)
    at Nimma.eval (eval at query (/Users/me/Projects/proj/node_modules/nimma/dist/legacy/cjs/core/index.js:64:270), <anonymous>:324:11)
    at Nimma.query (/Users/me/Projects/proj/node_modules/nimma/dist/legacy/cjs/core/index.js:66:71)
    at execute (/Users/me/Projects/proj/node_modules/@stoplight/spectral-core/dist/runner/runner.js:88:11)
    at Runner.run (/Users/me/Projects/proj/node_modules/@stoplight/spectral-core/dist/runner/runner.js:52:13)
    at Spectral.runWithResolved (/Users/me/Projects/proj/node_modules/@stoplight/spectral-core/dist/spectral.js:60:22)
    at async Spectral.run (/Users/me/Projects/proj/node_modules/@stoplight/spectral-core/dist/spectral.js:68:17)
    at async processInput (/Users/me/Projects/proj/node_modules/ibm-openapi-validator/src/cli-validator/run-validator.js:259:25)
```

The only divergence from the docs I have done is install it locally as a devDep rather than globally (`-g` flag).

I have verified that I do not have a global copy of the dependency on my machine.

Platform info:
- OS: macOS 11.5.2
- shell: zsh 5.8
- Node: 14.15.0
- npm: 8.3.0

Here is a link to my API definition:
https://github.com/loucadufault/Shopify-Backend-Developer-Intern-Challenge/blob/main/inventory-tracking-backend/docs/openapi.json

**Comment by dpopp07** [comment]
Hi @loucadufault sorry for the delay on this - I will do some debugging here today

**Comment by dpopp07** [comment]
Okay @loucadufault after throwing some debugging console log statements in the `nimma` code, I found the source of the error. It seems that Spectral/nimma is choking on [the parameterized server URL](https://github.com/loucadufault/Shopify-Backend-Developer-Intern-Challenge/blob/main/inventory-tracking-backend/docs/openapi.json#L14) in your API definition. It seems to be trying to create a Node `URL` object and is failing with the error:

`TypeError [ERR_INVALID_URL]: Invalid URL: http://127.0.0.1:{port}`

The `nimma` error looks like:

`RuntimeError: $ threw: NodeError("Invalid URL: http://127.0.0.1:{port}")`

The thing is, your API definition is definitely valid. Defining a variable in a URL like that is allowed and I don't know why Spectral wouldn't support it. Unfortunately, this is a problem with a dependency and not our tool so there's not a lot I can do but I'll open up an issue against Spectral describing the problem.

In the meantime, the only workaround is to modify your API definition to remove the variable. This will allow you to validate the document, then you can add it back later if need be.

**Comment by loucadufault** [comment]
Thanks for investigating this. It seems that the error occurs when the curly braces are part of the host port (in your example `http://some-service.io/v1:{port}`, the `:{port}` string is considered part of the pathname `v1:{port}`).

I am not sure I agree with your assessment that this is not an issue with Spectral. In my view, I wouldn't expect the string value for the Server Object `url` field to necessarily be parseable as a valid URL. Rather, I would expect that the variable placeholders get substituted by their values before parsing it as a URL and performing validation. Without having looked at the `nimma` code, I would argue that an implementation relying on parsing the "raw" template strings is erroneous. Is it possible there is some bug preventing the variable substitution from being properly performed before parsing the string?

Indeed, the very example showing "how variables can be used for a server configuration" over on the [OpenAPI Specification](https://swagger.io/specification/) (`https://{username}.gigantic-server.com:{port}/{basePath}`) is also not considered a valid URL by Node.js.

I am not familiar with the particulars of the [WHATWG URL Standard](https://url.spec.whatwg.org/) used by Node.js' `URL` constructor/parser, but in general curly braces are considered *unwise* characters to include in a URI according to [RFC 2396](https://www.ietf.org/rfc/rfc2396.txt). So it seems brittle that the implementation would rely on parsing a string intentionally containing these disallowed curly brackets (basically hoping that the template placeholders are such that the string still happens to form a valid URL according the the spec followed by the underlying Node.js parser).

**Comment by dpopp07** [comment]
> I am not sure I agree with your assessment that this is not an issue with Spectral. In my view, I wouldn't expect the string value for the Server Object url field to necessarily be parseable as a valid URL

Okay yes, I do agree with you. I meant that it did not seem to be a _bug_ with Spectral per se but intentional behavior that is indeed questionable. It is a bug though because it is an issue that Spectral, a tool that processes OpenAPI definitions, cannot process URLs clearly allowed in an OpenAPI definition. 

> I would argue that an implementation relying on parsing the "raw" template strings is erroneous

Also agreed. They don't seem to be resolving the URL variables before checking to see if they are valid. I doubt there is a bug preventing it, my guess is that they just aren't attempting to resolve (I could be wrong, I haven't looked at the code either).

I'll open an issue with Spectral about this. In the meantime, you'll have to temporarily edit your API definition to use the validator

**Comment by dpopp07** [comment]
Issue opened against Spectral: https://github.com/stoplightio/spectral/issues/2064

**Comment by loucadufault** [comment]
> I meant that it did not seem to be a bug with Spectral per se but intentional behavior that is indeed questionable.

Not a bug but a feature ;) I understand better now the intent of your previous comment.

> I'll open an issue with Spectral about this.

Appreciate it, and thanks for your support with this issue.

**Comment by dpopp07** [comment]
> Not a bug but a feature ;)

Haha yes, exactly

> Appreciate it, and thanks for your support with this issue.

No problem! Thanks for pointing this out, it's a good find. We'll see what the Spectral folks think about it

**Comment by dpopp07** [comment]
@loucadufault I was able to track this down to a custom rule within our ruleset (not Spectral). Not sure how that eluded me the first time around, but I'll try to put out a fix for this soon.

**Comment by dpopp07** [comment]
@loucadufault This is fixed in v0.58.2

**Comment by loucadufault** [comment]
Thanks for following up with the fix! Appreciate it.
